### ISSUE TITLE
GoogleAdsLoggingInterceptor Issue

### ISSUE THREAD

**Issue by koreansu** [googleads/google-ads-php#549]
I may just be missing something but I am getting started with the using the Google Ad API and am running into weird errors.  

I am seeing PHP Warnings that look as follows:

PHP Warning:  Declaration of Google\Ads\GoogleAds\Lib\V6\GoogleAdsLoggingInterceptor::interceptUnaryUnary($method, $argument, $deserialize, array $metadata, array $options, $continuation) should be compatible with Grpc\Interceptor::interceptUnaryUnary($method, $argument, $deserialize, $continuation, array $metadata = Array, array $options = Array) 

Then ultimately a Fatal error:
Fatal error: Uncaught TypeError: Argument 4 passed to Google\Ads\GoogleAds\Lib\V6\GoogleAdsLoggingInterceptor::interceptUnaryUnary() must be of the type array, object given, called in /home/vagrant/Code/datacollector/vendor/grpc/grpc/src/lib/BaseStub.php on line 423 and defined in /home/vagrant/Code/datacollector/vendor/googleads/google-ads-php/src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingInterceptor.php:53

I have installed the gRPC v 1.36 php extension. I can also confirm the grpc composer package is 1.36.  When I go to the package in my vendor folder and change the parameter ordering of GoogleAdsLoggingInterceptor::interceptUnaryUnary() to match 
Grpc\Interceptor::interceptUnaryUnary the error and warnings go away.  Going through the history of both packages, it seems like the code has always been this way. I'm stumped. I am trying to run the basic GetExpandedTextAds.php file.

**Comment by PierrickVoulet** [comment]
This issue is due to a breaking change introduced with gRPC 1.36.0 (this is similar to #541). We merged a fix #530 which will be live in the next release of our library.

In the meantime, you could either:
- Continue using gRPC 1.36.0 but patch the our library with our fix.
- Downgrade both the gRPC composer package and the gRPC PHP extension to 1.35.0.

**Comment by koreansu** [comment]
I'll be on the lookout for that new release and do one of those 2 options in the meantime. Thanks!

**Comment by PierrickVoulet** [comment]
@koreansu no ETA for the next release as of today.

**Comment by paslandau** [comment]
Just ran into the same issue after running `composer update`. Fixed it by manually adding `"grpc/grpc": "~1.35.0",` to our `composer.json` file to restrict the version.

@PierrickVoulet shouldn't that be something that you guys add to the composer.json of this library and release a patch version? Otherwise all your downstream users will eventually run into the same issue.

**Comment by PierrickVoulet** [comment]
@paslandau We are working on the next release and we hope that it will not take much longer. Anyone reaching the issue should downgrade gRPC to the version `1.35.0` in their application (both with composer **and** PECL).

**Comment by SeriousJ55** [comment]
I tried to fix it by manually adding "grpc/grpc": "~1.35.0", to our composer.json file, as recommended by paslandau. I still get the same error message. What now?

**Comment by paslandau** [comment]
- verify with "composer show" that you actually have the old version
```
www-data:/var/www/current# composer show | grep grpc/grpc
grpc/grpc                                      1.35.0             gRPC library for PHP
```
- check the offending file `\Grpc\Interceptor::interceptUnaryUnary`. The method SHOULD look likes this (order signature parameters is important!):
```php
    public function interceptUnaryUnary(
        $method,
        $argument,
        $deserialize,
        array $metadata = [],
        array $options = [],
        $continuation
    ) {
        return $continuation($method, $argument, $deserialize, $metadata, $options);
    }
```
- if you are using the grpc extension, make sure it also is in version `1.35.0`

**Comment by koreansu** [comment]
You could also implement the changes here: https://github.com/googleads/google-ads-php/pull/530/commits/02bcc6e908fbc7e4696e03741b323f6dd517746b

Definitely not ideal, but it does work. I figured a new release would have them anyway and I was struggling downgrading my env so this at least got me going.

**Comment by SeriousJ55** [comment]
I really don't want to spend time to fix again your faulty code. Is there a way to use previous SDK version and call API V6 through it? For example by simply copying the /V6/ directories?

**Comment by PierrickVoulet** [comment]
> I tried to fix it by manually adding "grpc/grpc": "~1.35.0", to our composer.json file, as recommended by paslandau. I still get the same error message. What now?

Do you use and if you do did you downgrade your PHP Extension for gRPC as well (`sudo pecl install grpc-1.35.0`)?

**Comment by SeriousJ55** [comment]
@paslandau:

- I verified with "composer show" and I actually have the old version 1.35.0
- I found many functions interceptUnaryUnary, all with parameters in the wrong order. I changed them all.

@paslandau and @PierrickVoulet:

- I don't know if I use gRPC. I can see in my phpinfo that I have gRPC 1.21.2 on my Linux server and gRPC 1.20.0 on my Wamp local env.

Before I try and change my gRPC, I ran the code with the corrected versions of functions interceptUnaryUnary and I ran into a different issue:

> Warning: require([path]\google-ads-php-master\vendor\composer/../phpunit/phpunit/src/Framework/Assert/Functions.php): failed to open stream: No such file or directory in [path]\google-ads-php-master\vendor\composer\autoload_real.php on line 66

Problem is I couldn't even find a phpunit directory. Don't you guys love it when trying to solve an issue brings many other different issues?

**Comment by PierrickVoulet** [comment]
@SeriousJ55 It sounds like you are using one version in composer (PHP package) but have another version installed on the environment (PHP Extension). You will need to fix that to be consistent if this is actually the case. The best would be to have the version `1.35.0`.

**Comment by SeriousJ55** [comment]
@paslandau:

- I checked grpc version with "composer show" and it is version 1.35.0
- I found many functions interceptUnaryUnary with parameters not in order. I changed them all back.

Then I ran my script again. This time I couldn't intercept any error message. The script just crashes with the generic error "Failed to read FastCGI header," which doesn't mean anything.

@paslandau and @PierrickVoulet:

- I never used grpc. I don't even know what it is. Maybe some of my other libraries use it in the background? I can see in my phpinfo that I have version 1.21.2 on my Linux server and version 1.20.0 on my Wamp server. Is there not a simpler way to make this terrible library work than to tinker with my environments?

**Comment by paslandau** [comment]
@SeriousJ55  I would probably just delete the whole vendor directory and run a "clean" composer install. Sounds like some of your dependencies got messed up in the process.

FYI: There shouldn't be any need to change files in the vendor folder to solve this issue.

**Comment by SeriousJ55** [comment]
@paslandau thank you for your help.

Unfortunately, I did run a clean composer install several times. It didn't work.

I just upgraded to grpc 1.35.0 on my Wamp local environment. Now it works. I'm not sure I'll be able to upgrage grpc on my Linux production server.

**Comment by PierrickVoulet** [comment]
@SeriousJ55 If you are not sure what is gRPC, feel free to read our dedicated [documentation](https://developers.google.com/google-ads/api/docs/client-libs/php/transport). To get started with this library, you can also take a look at the [README.md](https://github.com/googleads/google-ads-php/blob/master/README.md) file of the library, the sections [Requirements](https://github.com/googleads/google-ads-php/blob/master/README.md#requirements) and [Getting started](https://github.com/googleads/google-ads-php/blob/master/README.md#getting-started) should help.

**Comment by SeriousJ55** [comment]
@PierrickVoulet wouldn't it be better to make an SDK that works?

**Comment by SeriousJ55** [comment]
@paslandau: I finally managed to upgrade to grpc 1.35.0 on my production Linux server and it works (except for the fact that there's [no backward compatibility](https://github.com/googleads/google-ads-php/issues/558#event-4465834699)). Thank you for your help and good luck to those poor souls out there trying to figure out this SDK.

**Comment by PierrickVoulet** [comment]
We just released the version `v8.0.0`: It fixes this issue.
